### PR TITLE
chore(buffering): Handle buffering messages sequentially and add more docs 

### PR DESCRIPTION
**Pull request by olksdr**
This is just small fixes and mostly adding more docs about how sqlite behaves. 

This also should makes sure we handle all the incoming messages to the buffer service sequentially, which makes sure that we always have `write` before `read`.  

related to https://github.com/getsentry/team-ingest/issues/53

#skip-changelog

### CODE REVIEW

File: relay-server/src/actors/project_buffer.rs
Comment (jan-auer) [inline relay-server/src/actors/project_buffer.rs]: Shouldn't you exit early if begin fails? I'd refactor this to an async function that returns an error and then catch it in the spawn to log errors.

File: relay-server/src/actors/project_buffer.rs
Comment (jan-auer) [inline relay-server/src/actors/project_buffer.rs]: Where does it do that blocking?

Comment (olksdr) [reply relay-server/src/actors/project_buffer.rs]: Another query, which will try to access this table will be blocked till the lock is released. This is also partially address the `fence` suggestion from yesterday. There will be definitely only 1 query at a time.

File: relay-server/src/actors/project_buffer.rs
Comment (iker-barriocanal) [inline relay-server/src/actors/project_buffer.rs]: I'm not sure if I understand this correctly; but if I do, this comment is not correct. Since we're using the [WAL journal mode](https://www.sqlite.org/wal.html#how_wal_works), a `COMMIT` will [add a checkpoint](https://www.sqlite.org/wal.html#automatic_checkpoint) (so committing the changes to disk?). Then, indeed, the default committing policy is again in-place.

**Review by iker-barriocanal — COMMENTED** [review]
Other than the comment below, the rest looks good to me.

File: relay-server/src/actors/project_buffer.rs
Comment (olksdr) [reply relay-server/src/actors/project_buffer.rs]: I do not really understand your comment here. 
The comment for this is taken from the docs, just to make sure to clarify what's happening while we call `commit` on the open transaction.

File: relay-server/src/actors/project_buffer.rs
Comment (jjbayer) [inline relay-server/src/actors/project_buffer.rs]:
```suggestion
        // here to prepare the chunks which will be preparing the batch inserts.
```

File: relay-server/src/actors/project_buffer.rs
Comment (jjbayer) [inline relay-server/src/actors/project_buffer.rs:315]:
```suggestion
    /// It will spool to disk only if the persistent storage enabled in the configuration
    /// and a db connection can be acquired.
```

File: relay-server/src/actors/project_buffer.rs
Comment (jjbayer) [inline relay-server/src/actors/project_buffer.rs]: It looks like `sqlx` has a RAII wrapper for this: https://docs.rs/sqlx/latest/sqlx/struct.Transaction.html

Not sure if you want the automatic rollback in failure cases though.

File: relay-server/src/actors/project_buffer.rs
Comment (jjbayer) [inline relay-server/src/actors/project_buffer.rs]:
```suggestion
            // Removing envelopes from the on-disk buffer in batches has following implications:
```

File: relay-server/src/actors/project_buffer.rs
Comment (jjbayer) [inline relay-server/src/actors/project_buffer.rs]: Is it ok to do nothing if the connection cannot be acquired?

File: relay-server/src/actors/project_buffer.rs
Comment (jan-auer) [inline relay-server/src/actors/project_buffer.rs]: As a follow-up, we should: 
- try to spool as much as possible to fill up the disk (spool as many envelopes as possible)
- potentially: keep the rest in memory, and let the upstream service handle the case that we can no longer buffer. this must not "block" the caller on pending spooling, though.
- propagate an error up, via healthcheck and as 503 on queueing in the common endpoint handler.

File: relay-server/src/actors/project_buffer.rs
Comment (jan-auer) [inline relay-server/src/actors/project_buffer.rs]: From offline conversation: Intention here was to not use itertools. However, using some async function on a synchronous iterator from the `futures` crate doesn't buy us anything. Let's use itertools here.

File: relay-server/src/actors/project_buffer.rs
Comment (jan-auer) [inline relay-server/src/actors/project_buffer.rs]: This loop can take quite a while, and we're not yielding. This means that the thread on the main runtime that runs this future is going to be blocked for a long time. Should we either yield here, or defer this into a `spawn_blocking`? The latter is semantically wrong, and the CPU time has to be spent, but it should be spent in a context where we do not block other futures for too long.

Example:

```rust
let mut envelopes = Vec::new();
for k, vals in buffer {
  for val in vals {
    envelopes.insert((k, val.into_envelope().to_vec()));
    tokio::task::yield_now().await;  // TODO: consider doing this every Nth iteration
  }
}
```

File: relay-server/src/actors/project_buffer.rs
Comment (jan-auer) [inline relay-server/src/actors/project_buffer.rs:394]: From offline conversation: Provided this could be a DELTE without subquery if a specific feature is enabled, let's add that dependency ourselves, enable the feature, and leave a comment in cargo toml, why we have it.

File: relay-server/src/actors/project_buffer.rs
Comment (jan-auer) [inline relay-server/src/actors/project_buffer.rs]: We drop this service mostly when the runtime stops, which means we should no longer start async tasks. Also, our expectation would be that when we drop the service, we close the DB and it's connections, which now isn't true because the DB lives on in a spawned task. 

This kind of procedure should probably be implemented in response to the `shutdown` signal.

For the MVP, it's enough if we drop the buffer and log outcomes + errors.

**Review by jan-auer — APPROVED** [review]
Approved provided that open comments are addressed. Thanks!

File: relay-server/src/actors/project_buffer.rs
Comment (iker-barriocanal) [reply relay-server/src/actors/project_buffer.rs]: Do you mind linking to these docs then? Because I'm misunderstanding something.

File: relay-server/src/actors/project_buffer.rs
Comment (iker-barriocanal) [inline relay-server/src/actors/project_buffer.rs:239]: I also find the following relevant, from the [docs](https://www.sqlite.org/pragma.html#:~:text=A%20transaction%20committed%20in%20WAL%20mode%20with%20synchronous%3DNORMAL%20might%20roll%20back%20following%20a%20power%20loss%20or%20system%20crash.):

> A transaction committed in WAL mode with synchronous=NORMAL might roll back following a power loss or system crash.

File: relay-server/src/actors/project_buffer.rs
Comment (iker-barriocanal) [reply relay-server/src/actors/project_buffer.rs]: IMO `try_spool` should return an error in that case. If I call `try_spool` and it doesn't return an error, I'd expect I spooled. If there's any inner detail (i.e. connection could not be acquired), I'd expect I get the error so that I can choose to retry again if I want to.

File: relay-server/src/actors/project_buffer.rs
Comment (iker-barriocanal) [reply relay-server/src/actors/project_buffer.rs]: (also adding "the":)
```suggestion
            // Removing envelopes from the on-disk buffer in batches has the following implications:
```

File: relay-server/src/actors/project_buffer.rs
Comment (iker-barriocanal) [inline relay-server/src/actors/project_buffer.rs]: nit: I'm not sure if I like the "on-disk" here since this in the end is an abstraction, but let's move forward with this for now.

**Review by iker-barriocanal — APPROVED** [review]

File: relay-server/src/actors/project_buffer.rs
Comment (olksdr) [reply relay-server/src/actors/project_buffer.rs]: https://www.sqlite.org/lockingv3.html, see transaction control.

File: relay-server/src/actors/project_buffer.rs
Comment (olksdr) [reply relay-server/src/actors/project_buffer.rs]: This explicit open/close transaction has to go away now. And we will rely on the pool to do the right thing, since we handle all the messages sequentially now

File: relay-server/src/actors/project_buffer.rs
Comment (olksdr) [reply relay-server/src/actors/project_buffer.rs]: As agreed, adding this to followups and will jump one some of the items after merging this PR.

File: relay-server/src/actors/project_buffer.rs
Comment (olksdr) [reply relay-server/src/actors/project_buffer.rs:394]: This is the related issue https://github.com/rusqlite/rusqlite/issues/1111, which explains the problem with this. 
If we want to enable this feature, we will have to build `sqlite` on our own. There is no easy way to just enable the feature.

File: relay-server/src/actors/project_buffer.rs
Comment (iker-barriocanal) [reply relay-server/src/actors/project_buffer.rs]: Thanks! I misunderstood what checkpointing was, thus my confusion. It makes sense.

**Pull request merged**
